log/uds.go

```diff
@@ -0,0 +1,109 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package log
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/cenkalti/backoff"
+	"go.uber.org/zap/zapcore"
+)
+
+// An udsCore write entries to an UDS server
+type udsCore struct {
+	client       http.Client
+	minimumLevel zapcore.Level
+	url          string
+	maxAttempts  int
+}
+
+// teeToUDSServer returns a zapcore.Core that writes entries to both the provided core and to an uds server.
+func teeToUDSServer(baseCore zapcore.Core, address, path string, maxRetryAttempts int) zapcore.Core {
+	c := http.Client{
+		Transport: &http.Transport{
+			DialContext: func(_ context.Context, _, _ string) (net.Conn, error) {
+				return net.Dial("unix", address)
+			},
+		},
+		Timeout: 100 * time.Millisecond,
+	}
+	uc := &udsCore{client: c, url: "http://unix" + path, maxAttempts: maxRetryAttempts + 1}
+	for l := zapcore.DebugLevel; l <= zapcore.FatalLevel; l++ {
+		if baseCore.Enabled(l) {
+			uc.minimumLevel = l
+			break
+		}
+	}
+	return zapcore.NewTee(baseCore, uc)
+}
+
+// Enabled implements zapcore.Core.
+func (u *udsCore) Enabled(l zapcore.Level) bool {
+	return l >= u.minimumLevel
+}
+
+// With implements zapcore.Core.
+func (u *udsCore) With(fields []zapcore.Field) zapcore.Core {
+	return &udsCore{
+		client:       u.client,
+		minimumLevel: u.minimumLevel,
+	}
+}
+
+// Check implements zapcore.Core.
+func (u *udsCore) Check(e zapcore.Entry, ce *zapcore.CheckedEntry) *zapcore.CheckedEntry {
+	if u.Enabled(e.Level) {
+		return ce.AddCore(e, u)
+	}
+	return ce
+}
+
+// Sync implements zapcore.Core.
+func (u *udsCore) Sync() error {
+	return nil
+}
+
+// Write implements zapcore.Core. It writes a log entry to an UDS server.
+func (u *udsCore) Write(entry zapcore.Entry, fields []zapcore.Field) error {
+	attempts := 0
+	b := backoff.NewExponentialBackOff()
+	var lastError error
+	for attempts < u.maxAttempts {
+		attempts++
+		resp, err := u.client.Post(u.url, "text/plain", strings.NewReader(entry.Message))
```

Comment:
HTTP for logs seems slightly suspicious. There is definitely some prior art here with k8s/containers - is there a simpler approach?

This is probably fine since we have low load so I wouldn't change it for something more complex though

Reply:
> There is definitely some prior art here with k8s/containers - is there a simpler approach?

If you are talking about forwarding container log to storage backend, like fluentd, I think they deal with log files. However it is bit complicated with CNI plugin, because there can be multiple CNI plugin process running simultaneously and write to the same file, which is not handled by file rotation library like https://github.com/bdurand/lumberjack, and log file read at CNI container is also going to be a bit complicated considering file rotation. So feels to me using rpc is probably easier considering load is small.

Re the protocol, these are the protocols supported by fluent: https://docs.fluentd.org/input seems like all other protocols other than http are going to add quite some complexity, like customized codec on the log stream, or import some uncommon fluent golang library. Both are less ideal than just use plain http.

Reply:
Ok, seems reasonable to me

Reply:
is UDP not simpler and lower overhead?  WriteToUDP / ReadFromUDP()?

Reply:
> is UDP not simpler and lower overhead? WriteToUDP / ReadFromUDP()?

I'd rather not invent a new protocol for the log stream in order to use UDP. IMO HTTP performance is OK considering the low load.

Reply:
Do we need a protocol beyond write msg bytes and read msg bytes?

Should we add a bunch of documentation around the load, etc., to the option if we do use HTTP POSTs, so that non-CNI users understand the intent?

Reply:
> Do we need a protocol beyond write msg bytes and read msg bytes?

Yeah probably not, we could put everything into a JSON obj and pass it as a message. However, I am updating this to buffer & send, with that, connection overhead should not be considerable? HTTP vs something + UDP probably does not matter any more.

Reply:
OK.